Error if AppArmor policies are not loaded

ewindisch opened the pull request:
Because we now expect the packaging (or admin) to install
the AppArmor policy, we should check on loading the execdriver
if the policy is loaded. If it is not loaded, error.

Fixes #9059

Signed-off-by: Eric Windisch eric@windisch.us

ewindisch committed:
Error if AppArmor policies are not loaded

Because we now expect the packaging (or admin) to install
the AppArmor policy, we should check on loading the execdriver
if the policy is loaded. If it is not loaded, error.

Signed-off-by: Eric Windisch <eric@windisch.us>

ewindisch commented:
/cc @duglin @stefanberger @salmanbaset

ewindisch commented:
/cc @moxiegirl (are you okay with the note to a github url? Should we have a doc page for this?)

moxiegirl commented on daemon/execdriver/native/driver.go:
```diff
@@ -51,6 +52,12 @@ func NewDriver(root, initPath string, options []string) (*driver, error) {
 
+	if apparmor.IsEnabled() {
+		if err := hasAppArmorProfileLoaded("docker-default"); err != nil {
+			return nil, fmt.Errorf("AppArmor enabled on system, but docker-default profile not loaded. Install policies from contrib/apparmor (https://github.com/docker/docker/tree/master/contrib/apparmor)")
```
---> This might be shorter --- I don't think you lose anything by using the shortened URL

AppArmor is enabled but the docker-default profile is not loaded. Install policies from contrib/apparmor (https://goo.gl/CjGtgC) and try again."

ewindisch replied:
Thanks. I've updated it to the first option.

moxiegirl commented on daemon/execdriver/native/driver.go:
```diff
@@ -51,6 +52,12 @@ func NewDriver(root, initPath string, options []string) (*driver, error) {
 
+	if apparmor.IsEnabled() {
+		if err := hasAppArmorProfileLoaded("docker-default"); err != nil {
+			return nil, fmt.Errorf("AppArmor enabled on system but the docker-default profile is not loaded. Install policies from contrib/apparmor (https://goo.gl/CjGtgC) and try again.")
```
@ewindisch one of us forgot the is in "AppArmor enabled"  which should be "AppArmor is enabled"   --- minor nit. If you don't want to kill it.  I  understand. Otherwise LGTM

ewindisch committed:
Documentation for AppArmor

Provide documentation for using the AppArmor
policies. Errors in AppArmor may point to this
documentation which is intentionally colocated
with the policies themselves to simplify installation.

Signed-off-by: Eric Windisch <eric@windisch.us>

ewindisch commented:
@moxiegirl thanks. I've also added a README.md here!

moxiegirl commented:
@ewindisch Cool. Will we ever need to put the info from the README in the user-facing docs?

ewindisch commented:
@moxiegirl I actually found that this breaks with the docs in the contrib/apparmor directory so we'll need to move this somewhere smarter underneath docs/. Suggestion?

ewindisch commented:
Moved the docs to docs/security/apparmor.md... although it's not linked from anywhere (yet)

moxiegirl commented:
@ewindisch Apologies for the slow reply.  Yes the move looks good. Now that this is in the docs. I'll review for that location.  I'll get it done today.

moxiegirl commented on docs/security/apparmor.md:
```diff
@@ -0,0 +1,87 @@
+
+If AppArmor is enabled on the host, Docker will expect
+the `docker-default` and `docker-unconfined` policies
+to exist and will apply them automatically when starting
```
To exist where? Is there a well known location?

ewindisch replied:
will except these policies to be loaded.  I can revise.

moxiegirl replied:
@ewindisch  If you can start with the DRAFT I provided that would be great.  Otherwise, I can take a stab at it.  Let me know which you prefer

moxiegirl commented on docs/security/apparmor.md:
````diff
@@ -0,0 +1,87 @@
+For example:
+
+```
+$ sudo apparmor_parser -r contrib/apparmor
````
Is this an example of loading a policy?  

This page uses a lot of words "installing" "loading" and "applying" policies. Are these three different actions and are they related? Or is this a case of three different words that all really refer to the same action

ewindisch replied:
Yes, this is loading a policy.

The difference between loading and installing is that installing is persisting across reboots.

Applying a policy is *using* a policy that has been loaded.

moxiegirl replied:
Ah, nicely said. We should say that here too.

moxiegirl commented on docs/security/apparmor.md:
```diff
@@ -0,0 +1,87 @@
+AppArmor Policies for Docker
+----------------------------
+
+This directory contains two files and three policies for Docker:
```
"This directory" is ???

ewindisch replied:
The directory is contrib/apparmor. I moved the file!

moxiegirl commented:
@ewindisch So, I have some unanswered questions regarding this facility.  It looks like not every Linux distro is impacted. It also appears only distros that are building their own packages or otherwise futzing about deving stuff need to worry about doing this.

I'd like to see some examples of these policies...I don't see any here but I haven't looked at installing docker on a Debian system. I'll check that in the AM.  I also think if we are going to talk about these things in the docs we should provide a bit more guidance/info about them.

Here is my **ROUGH CUT** on this. You'll see I have some unanswered questions.  If installing a Debian instance is the way to see the actual policies, I'll do that on Monday to confirm what I am seeing.

--->

<!--[metadata]>
+++
title = "AppArmor security policies for Docker"
description = "AppArmor security policies for Docker"
keywords = ["docker, documentation, about, technology, apparmor"]
[menu.main]
parent = "mn_use_docker"
+++
<![end-metadata]-->

# AppArmor security policies for Docker

AppArmor (Application Armor) is a security module that allows a system administrator to associate a security profile with each program. Docker expects to find an AppArmor policy loaded and enforced. 

Typically, AppArmor policies are installed by the `.deb` packages for
Debian-derived distributions, including Ubuntu. However, if you are building your own packages or developing Docker, you may encounter errors such as: 

```
Error response from daemon: Cannot start container 41398c949d5e4eae15796f651cf87359a51499e3861f98779e8f326eadc08235: set apparmor profile docker-default: no such file or directory
```

This error message indicates AppArmor polices are not enabled. This section explains how to manually install AppArmor policies. It also explains how to override a policy for a specific container instance.

>**Note**: Other packages are free to incorporate these policies
and operating system vendors are encouraged to install policies with Docker.

## Understand the policies

The XXXX directory contains the `docker` and `docker-engine` policy files.  The `docker` defines both the `docker-default` and  the `docker-unconfimed` policy. 
While the `docker-engine` policy defines the Docker Engine policy which applies to the `docker` client and daemon.

The `docker-default` policy is the default for running containers. It is a
moderately protective policy that provides wide application compatability. Use this policy if you XXX_SUPPLY_REASON_HERE.

The `docker-unconfined` policy is a replacement for a true 'unconfined' policy
that executes processes within a child context. This is the policy used by
default when specifying a privileged container. Users of privileged containers
may still override this policy on a per-container basis.


The Docker Engine policy is simply called `/usr/bin/docker` and is
apply automatically when loaded into AppArmor. This policy applies to
both the client and the daemon. The policy enforces a principle of least-privilege for the daemon. This principle means XXXX. 

## Installing Policys

If AppArmor is enabled on the host, Docker will expect
the `docker-default` and `docker-unconfined` policies
to exist and will apply them automatically when starting
containers.

For example:

```
$ sudo apparmor_parser -r contrib/apparmor
```

Refer to your operating system's documentation on
persisting these policies across reboots.


## Override the policy for a container


Users may override the AppArmor profile using the `security-opt` option
(per-container).  To use your system's standard `unconfined` policy, explicitly
set the security-opt, `apparmor:unconfined`. This policy will inherit all
system-wide policies, applying path-based policies intended for the host system
inside of containers. This was the default for privileged containers prior to
Docker 1.8.

For example, the following explicitly specifies the default policy:

```
$ docker run --rm -it --security-opt apparmor:docker-default hello-world
```

jfrazelle commented on daemon/execdriver/native/driver.go:
```diff
@@ -51,6 +52,15 @@ func NewDriver(root, initPath string, options []string) (*driver, error) {
+		apparmor_policies := []string{"docker-default", "docker-unconfined"}
+		for _, policy := range apparmor_policies {
+			if err := hasAppArmorProfileLoaded(policy); err != nil {
+				return nil, fmt.Errorf("AppArmor enabled on system but the %s profile is not loaded. Install policies from contrib/apparmor (https://goo.gl/CjGtgC) and try again.", policy)
```
I think it should be a warning to begin with so we dont break people

ewindisch replied:
The problem is that if we don't error here it *does* break people. This is why the docker-unconfined patch was reverted. The policy didn't exist and the daemon just moved on, then broken. Similarly, if docker-default isn't there, the daemon would similarly break.

The alternative for 'docker-unconfined' is that if it doesn't exist, that (for one release) we could fallback to 'unconfined'. Although that might cause more harm, especially if we're okay with the installation / usage docs (and packages).

ewindisch commented:
closing for #15099
